## Linear Programming Solver 

### Object Allows User to Call & Enter Values for Standard Form to Solve Linear Optimization Problems

* Contructor: Asks User for Input
* Simplex(): Solves Optimization Problem, Outputs Matrix for Result Interpretation

In [220]:
import numpy as np

In [217]:
class LinearProgramming(object):

    def __init__(self):

        print("Welcome to Linear Programming Solver")
        print("Please Enter the Details of Your Linear Program Below")

        print("")
        self.numOfVariables=int(input("Please Enter Number of Inputs: "))
        self.numOfConstraints=int(input("Please Enter Number of Constraints: "))


        self.matrix= np.zeros((self.numOfConstraints+1, self.numOfVariables+self.numOfConstraints+1+1))

      
        self.row=self.numOfConstraints+1
        self.col=self.numOfVariables+self.numOfConstraints+1+1

        print("")
        print("Enter Variables for Maximization Function")
        for i in range(self.numOfVariables):
            variable=float(input("Value for x{}: ".format(i+1)))
            self.matrix[self.row-1][i]=-variable
            
        self.matrix[self.row-1][self.col-2]=1
        print("")
        print("Enter Variables for Constraint Functions")

        for j in range(self.numOfConstraints):
            
            print("Constraint # {} ".format(j+1))
            for i in range(self.numOfVariables):
                variable=float(input("Value for x{}: ".format(i+1)))
                self.matrix[j][i]=variable

            variable=float(input("Value for <= "))
            self.matrix[j][self.col-1]=variable
            self.matrix[j][self.numOfVariables+j]=1
            print("")


    def reduceRows(self,rowNum,colNum,row):
        
        for i in range(self.row):
            if i!=rowNum:
           
                original_row=np.array(self.matrix[i],copy=True)
               
                mult=original_row[colNum]/row[colNum]
               
                
                for j in range(len(original_row)):
                 
                    original_row[j]=original_row[j]-(row[j]*mult)
                
                
                self.matrix[i]=original_row
                

    def pivotIndex(self,index):
        col=self.matrix[:,self.col-1]
        col[:self.row-1]
       
        list=[]
        
        for i in range(len(col[:self.row-1])):
            if self.matrix[:,index][i]>0:
                list.append(col[i]/self.matrix[:,index][i])
            else:
                list.append(float("inf"))
        
       
        
        self.matrix[list.index(min(list))]=self.matrix[list.index(min(list))]/self.matrix[:,index][list.index(min(list))]
        
        self.reduceRows(list.index(min(list)),index,self.matrix[list.index(min(list))])
        
        
        
    def Simplex(self):
        
        count=0
        while(1):
            
            row=self.matrix[self.row-1]

            minimun=min(row[:self.col-1])

            if minimun<0:
                index=np.where(row==min(row[:self.col-1]))[0][0]  
                self.pivotIndex(index)
                count=count+1
                
               
            else:

                break


## The Blending Problem

### A Simple Linear Programming Problem, Inputs are Assumed to be in Standard Form


#### Maximize:

* 20x1 + 50x2

#### Constraints:

* -0.2x1 + 0.8x2 <= 0
* 2x1 + 4x2 <= 240
* 1x1 + 0x2 <= 100

In [218]:
blending=LinearProgramming()


Welcome to Linear Programming Solver
Please Enter the Details of Your Linear Program Below

Please Enter Number of Inputs: 2
Please Enter Number of Constraints: 3

Enter Variables for Maximization Function
Value for x1: 20
Value for x2: 50

Enter Variables for Constraint Functions
Constraint # 1 
Value for x1: -0.2
Value for x2: 0.8
Value for <= 0

Constraint # 2 
Value for x1: 2
Value for x2: 4
Value for <= 240

Constraint # 3 
Value for x1: 1
Value for x2: 0
Value for <= 100



In [219]:
blending.Simplex()
np.set_printoptions(suppress=True)
print(blending.matrix)

[[   0.       1.       0.833    0.083    0.       0.      20.   ]
 [   1.       0.      -1.667    0.333    0.       0.      80.   ]
 [   0.       0.       1.667   -0.333    1.       0.      20.   ]
 [   0.       0.       8.333   10.833    0.       1.    2600.   ]]


## Solution

* ### Maximum Profit is 2600 with x1=80, x2=20

## The Production Problem

### A Boat Manufacturer produces 2 types of boat, Boat 1 produces 50 in profit & Boat 2 produces 60


#### Needs:

* Boat 1: 50 lbs aluminum, 6 min of machine time, 3 hrs of labor
* Boat 2: 30 lbs aluminum, 5 min of machine time, 5 hrs of labor

#### Availability:
* Aluminum: 2000 lbs
* Machine Time: 300 min
* Labor: 200 hrs

production=LinearProgramming()

In [226]:
production.Simplex()
np.set_printoptions(suppress=True)
print(production.matrix)

[[   1.       0.       0.031    0.      -0.188    0.      25.   ]
 [   0.       0.      -0.094    1.      -0.438    0.      25.   ]
 [   0.       1.      -0.019    0.       0.312    0.      25.   ]
 [   0.       0.       0.438    0.       9.375    1.    2750.   ]]


## Solution

* ### Maximum Profit is 2750 with 25 Boat 1 Produced & 25 Boat 2 Produced
